stl/inc/expected

```diff
@@ -0,0 +1,1035 @@
+        if (_Has_value) {
+            _STD construct_at(_STD addressof(_Unexpected), _STD move(_Other._Unexpected));
+            _Has_value = false;
+        } else {
+            _Unexpected = _STD move(_Other._Unexpected);
+        }
+
+        return *this;
+    }
+
+    constexpr void emplace() noexcept {
+        if (!_Has_value) {
+            if constexpr (!is_trivially_destructible_v<_Err>) {
+                _Unexpected.~_Err();
+            }
+            _Has_value = true;
+        }
+    }
+
+    // [expected.void.swap]
+    constexpr void swap(expected& _Other) noexcept(
+        is_nothrow_move_constructible_v<_Err>&& is_nothrow_swappable_v<_Err>) //
+        requires is_swappable_v<_Err> && is_move_constructible_v<_Err> {
+        if (_Has_value && _Other._Has_value) {
+            // nothing
+        } else if (_Has_value) {
+            _STD construct_at(_STD addressof(_Unexpected), _STD move(_Other._Unexpected));
+            if constexpr (!is_trivially_destructible_v<_Err>) {
+                _Other._Unexpected.~_Err();
+            }
+            _Has_value        = false;
+            _Other._Has_value = true;
+        } else if (_Other._Has_value) {
+            _STD construct_at(_STD addressof(_Other._Unexpected), _STD move(_Unexpected));
+            if constexpr (!is_trivially_destructible_v<_Err>) {
+                _Unexpected.~_Err();
+            }
+            _Has_value        = true;
+            _Other._Has_value = false;
+        } else {
+            _Swap_adl(_Unexpected, _Other._Unexpected);
+        }
+    }
+
+    friend constexpr void swap(expected& _Left, expected& _Right) noexcept(
+        is_nothrow_move_constructible_v<_Err>&& is_nothrow_swappable_v<_Err>) //
+        requires is_swappable_v<_Err> && is_move_constructible_v<_Err> {
+        if (_Left._Has_value && _Right._Has_value) {
+            // nothing
+        } else if (_Left._Has_value) {
+            _STD construct_at(_STD addressof(_Left._Unexpected), _STD move(_Right._Unexpected));
+            if constexpr (!is_trivially_destructible_v<_Err>) {
+                _Right._Unexpected.~_Err();
+            }
+            _Left._Has_value  = false;
+            _Right._Has_value = true;
+        } else if (_Right._Has_value) {
+            _STD construct_at(_STD addressof(_Right._Unexpected), _STD move(_Left._Unexpected));
+            if constexpr (!is_trivially_destructible_v<_Err>) {
+                _Left._Unexpected.~_Err();
+            }
+            _Left._Has_value  = true;
+            _Right._Has_value = false;
+        } else {
+            _Swap_adl(_Left._Unexpected, _Right._Unexpected);
+        }
+    }
+
+    // [expected.void.obs]
+    _NODISCARD constexpr explicit operator bool() const noexcept {
+        return _Has_value;
+    }
+    _NODISCARD constexpr bool has_value() const noexcept {
+        return _Has_value;
+    }
+
+    constexpr void operator*() const noexcept {}

```

Comment:
Should we check the `has_value()` precondition here when `_CONTAINER_DEBUG_LEVEL > 0`? The behavior is well-defined anyway, but it might help people write more-generic code. I think I don't _actually_ care.

Reply:
Neutral here.